modules/ank/arrow-ank/src/main/kotlin/arrow/ank/interpreter.kt

```diff
@@ -210,46 +211,49 @@ fun <F> monadDeferInterpreter(MF: MonadDefer<F>): AnkOps<F> = object : AnkOps<F>
           }
         }
       )
-    }
+    })
 
-  override fun generateFile(path: Path, newContent: String): Kind<F, Unit> = MF.delay {
-    Files.write(path, newContent.toByteArray())
-    Unit
-  }
+  override fun generateFile(path: Path, newContent: Sequence<String>): Path =
+    Files.write(path, newContent.asIterable())
 
-  private fun createEngineCache(snippets: Nel<Snippet>, compilerArgs: List<String>): Kind<F, Map<String, ScriptEngine>> = MF.run {
-    bindingCatch {
-      val classLoader = delay { URLClassLoader(compilerArgs.map { URL(it) }.toTypedArray()) }.bind()
-      val seManager = delay { ScriptEngineManager(classLoader) }.bind()
-      delay {
-        snippets.all.asSequence().distinctBy { it.lang }.map {
-          it.lang to seManager.getEngineByExtension(extensionMappings.getOrDefault(it.lang, "kts"))
-        }.toMap()
-      }.bind()
+  private val engineCache: ConcurrentMap<List<String>, Map<String, ScriptEngine>> = ConcurrentHashMap()
+
+  private fun getEngineCache(snippets: Sequence<Snippet>, compilerArgs: List<String>): Map<String, ScriptEngine> {
+    val cache = engineCache[compilerArgs]
+    return if (cache == null) { //create a new engine
+      val classLoader = URLClassLoader(compilerArgs.map { URL(it) }.toTypedArray())
+      val seManager = ScriptEngineManager(classLoader)
+      val langs = snippets.map { it.lang }.distinct()
+      val engines = langs.toList().map {
+        it to seManager.getEngineByExtension(extensionMappings.getOrDefault(it, "kts"))
+      }.toMap()
+      engineCache.putIfAbsent(compilerArgs, engines) ?: engines
+    } else { //reset an engine. Non thread-safe
+      cache.forEach { _, engine ->
```

Comment:
*wince*

Reply:
This is the most performant way because it allows to cache the engine instances but they need to be reset. it's non-thread safe because ank by design is sequential to avoid minimal memory consumption since it has to do a fair amount of String manipulation. Open to suggestions.